### PR TITLE
[Tags] Add tiered roles

### LINKED ISSUE #292: [Tags] Allow tiered roles
Both as part of feedback from #286 and before, make roles configurable:
- Tiered roles for max number of tags per given role.
- Role to make add/rm commands visible.

One way that we could do this is to:
- [x] Modify `max` command to accept a role on top of max tags. Could be `max <role> <maxTags>`. This requires the backend config to change, so perhaps move this part to Config v3.
- [x] Modify `user_exceeds_tag_limit` to take into account the roles the member has, and take the max limit of all roles.
    - [x] Modify return value to include the maximum number of tags they can have.
- [x] Modify message to send to user when they exceed the limit to reflect the actual limit.

### PR DESCRIPTION
**Pull request by Injabie3**
### Type
- [x] Enhancement

### Description of the changes
Closes #292 . Add tag tiers, where each role can have a maximum number of tags. If a user has multiple roles that are tiered, then the user takes the maximum of those roles.

- Use Config V3 for storing the tier settings.
- Nest server settings commands in a new `settings` group, accessible via `[p]tag settings`.
- Add `tiers` command to show each role and their relevant max tags.
- Modify `max` command to accept both a role and maximum number of tags.
- Modify messages that are sent to the user if they are over the maximum to reflect their tiers.

### CODE REVIEW

File: cogs/tags/tags.py [hunk 1]
```diff
@@ -200,13 +205,14 @@ async def user_exceeds_tag_limit(self, server: discord.Guild, user: discord.Memb
 
         Returns:
         --------
-        bool
-            True if too many tags, else False.
+        (bool, int)
+            bool: True if user has too many tags, else False.
+            int: The maximum number of tags this user can have. -1 if they
```
Comment (IAmTheRedSpy) [inline cogs/tags/tags.py]: Returning infinity if the user has no limit would allow for simpler logic

File: cogs/tags/tags.py [hunk 2]
```diff
@@ -200,13 +205,14 @@ async def user_exceeds_tag_limit(self, server: discord.Guild, user: discord.Memb
...
+            have no limit.
         """
-
         if await self.bot.is_owner(user):
             # No limit for bot owner
```
Comment (IAmTheRedSpy) [inline cogs/tags/tags.py:214]: Minor, but if the owner wants to have a limit they can't

File: cogs/tags/tags.py [hunk 3]
```diff
@@ -200,13 +205,14 @@ async def user_exceeds_tag_limit(self, server: discord.Guild, user: discord.Memb
...
-            return False
+            return (False, NO_LIMIT)
 
         if server:
             # No limit for mods and admins of server.
```
Comment (IAmTheRedSpy) [inline cogs/tags/tags.py:218]: Same as for owner

File: cogs/tags/tags.py [hunk 4]
```diff
@@ -271,33 +283,65 @@ def verify_lookup(self, lookup):
...
+        role, then they will take the MAXIMUM number from the roles that they
+        have.
 
         This limit does not apply to admins or mods.
 
         Parameters:
         -----------
+        role: discord.Role
+            The role to set a maximum for.
         num_tags: int
```
Comment (IAmTheRedSpy) [inline cogs/tags/tags.py:307]: Add a way to allow unlimited tags for a role. Maybe using the keyword "infinity" or "unlimited"

Ex: An admin wants to make a "unlimited_tags" role and give it to a handful of users.

File: cogs/tags/tags.py [hunk 5]
```diff
@@ -271,33 +283,65 @@ def verify_lookup(self, lookup):
...
+        for roleId, maxTags in tiers.items():
+            role = discord.utils.get(ctx.guild.roles, id=int(roleId))
+            if not role:
+                continue
+            validTiers.append((role, maxTags))
+        if validTiers:
+            msg = "Below is a list of roles and their tag limits:\n"
+            validTiers.sort(key=lambda x: x[1])
+            for role, maxTags in validTiers:
+                msg += f"{role.name}: {maxTags}\n"
```
Comment (IAmTheRedSpy) [inline cogs/tags/tags.py]: Minor:
Add the strings to a list and then set msg to the concatenated list after the loop
```
msg = "".join(msg_list)
```

**Review by IAmTheRedSpy — CHANGES_REQUESTED** [review]

File: cogs/tags/tags.py (hunk 4 above)
Comment (Injabie3) [reply cogs/tags/tags.py:307]: Hmmm, what if I use the same `NO_LIMIT` constant, and special case that?

File: cogs/tags/tags.py (hunk 1 above)
Comment (Injabie3) [reply cogs/tags/tags.py]: I'll look into using `float('inf')`.

File: cogs/tags/tags.py (hunk 2 above)
Comment (Injabie3) [reply cogs/tags/tags.py:214]: Will disregard for now, since this exists already.

File: cogs/tags/tags.py (hunk 4 above)
Comment (IAmTheRedSpy) [reply cogs/tags/tags.py:307]: If NO_LIMIT is infinity then you can just set num_tags to NO_LIMIT

File: cogs/tags/tags.py (hunk 1 above)
Comment (Injabie3) [reply cogs/tags/tags.py]: Used `float("inf")` to replace `NO_LIMIT`.

File: cogs/tags/tags.py (hunk 4 above)
Comment (Injabie3) [reply cogs/tags/tags.py:307]: Sure, done.

File: cogs/tags/tags.py (hunk 5 above)
Comment (Injabie3) [reply cogs/tags/tags.py]: Sure, fixed.

**Review by IAmTheRedSpy — APPROVED** [review]
You could generalize user_exceeds_tag_limit by setting admin and owner roles to NO_LIMIT during init. That might be less efficient so I think it's fine the way it is.

**Comment by Injabie3** [comment]
Thanks for the review!

**Pull request merged**